File: Algorithms/072.EditDistance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int D[word1.size()+1][word2.size()+1];
        for (int i=0; i <= word1.size(); i++)
            D[i][0] = i;
        for (int i=0; i <= word2.size(); i++)
            D[0][i] = i;
        for (int i=1; i <= word1.size(); i++) {
            for (int j=1; j <= word2.size(); j++) {
                D[i][j] = min(D[i][j-1], D[i-1][j]) + 1;
                D[i][j] = min(D[i][j], D[i-1][j-1] + (word1[i-1] == word2[j-1]? 0:1));
            }
        }
        return D[word1.size()][word2.size()];
    }
};

File: Algorithms/py/072.EditDistance.py
class Solution:
    # @return an integer
    def minDistance(self, word1, word2):
        m, n = len(word1), len(word2)
        distance = [[0] * (n + 1) for x in range(m + 1)]
        for i in range(m + 1):
            distance[i][0] = i
        for i in range(n + 1):
            distance[0][i] = i
        for i in range(1, m+1):
            for j in range(1, n+1):
                distance[i][j] = min(distance[i][j-1], distance[i-1][j]) + 1
                distance[i][j] = min(distance[i][j],
                        distance[i-1][j-1] + (0 if word1[i-1] == word2[j-1] else 1))
        return distance[m][n]
